onnxruntime/core/optimizer/unsqueeze_elimination.h

```diff
@@ -3,16 +3,23 @@
 
 #pragma once
 
-#include "core/optimizer/graph_transformer.h"
+#include "core/optimizer/rewrite_rule.h"
 
 namespace onnxruntime {
 
-class UnsqueezeElimination : public onnxruntime::GraphTransformer {
+class UnsqueezeElimination : public RewriteRule {
  public:
-  UnsqueezeElimination() noexcept : onnxruntime::GraphTransformer("EliminateUnsqueeze", "Eliminate unsqueeze node") {}
+  UnsqueezeElimination() noexcept : RewriteRule("EliminateUnsqueeze", "Eliminate unsqueeze node") {}
 
  private:
-  Status ApplyImpl(onnxruntime::Graph& graph, bool& modified, int graph_level) const override;
+  /** Apply rule when op type is one of the following. */
+  const std::string included_op_type_ = "Unsqueeze";
+
+  bool SatisfyCondition(const Graph& graph, const Node& node) override;
+
+  bool OpTypeCondition(const Node& node) override;
```

Comment:
>OpTypeCondition [](start = 7, length = 15)

:nit do you need a separate OpTypeCondition? Can't this be a part of StatisfyCondition?

Plus it is only being called from SatisfyCondition

Reply:
I want it so that people know that there is always a check for the op type. We can do some refactoring eventually to make sure it is always called (and then other conditions are called too).
The check is different for each rule, but it should always be there.

Reply:
I understood that part... my point is since this is being called from SatisfyCondition there is no way to guarantee this will always be called... 

---
In reply to: [267895985](https://github.com/Microsoft/onnxruntime/pull/670#discussion_r267895985) [](ancestors = 267895985)

Reply:
That is correct. What I meant above regarding the refactoring is that once we have a few rules, we can find common pattern and have checks like the OpTypeCondition to be always applied. Then each rule subclass can add more to it. Having a check like the OpTypeCondition will allow such refactoring happen very easily.

Reply:
I have same idea as @askhade.  It is better to remove OpTypeCondition now.  Otherwise, every re-write rule needs to implement this method.

Reply:
I am trying to give some structure to the conditions.
Removing it is minimal work.
But I feel we are making it less and less structured with every commit, instead of the opposite.
This is just a condition. We can make it return true by default and let rules override it. But I don't think we even need to do that, given that all the rules we have in mind require an OpTypeCondition. 
My idea was to force users to always override it to make their rules more efficient. It should be the first condition to check as it is the one that will lead to quickly filtering out the nodes for which a specific rule cannot be applied (this was also the idea behind the hashmap we had before).
Like I said above, I can improve the current implementation so that we always call it first and then call other conditions.

Reply:
I added comments in the rewrite_rule.h to let developers know how they should implement their rules.
To make things a bit more clear with the conditions, SatisfyConditions now makes two checks: OpTypeCondition (should always come first to discard inapplicable rules fast) and AdditionalConditions (in which any other condition can be added). If later we see commonalities between conditions (other than the op type one), we can create a separate method.

Reply:
In my opinion, it is not a good idea to make rewrite rule API so complicate. Please re-consider to remove OpTypeCondition and AdditionalConditions.

Reply:
I am asking to override at most two methods: one to quickly check the op type (I make sure this is called first for quickly discarding unrelated rules) and one to add all other checks.
I cannot see why this is complicated.
With the latest implementation, they can implement any of the two they want and they also don't have to add those manually in the SatisfyCondition (unlike what I was doing before and was related to Ashwini's comment).
@skottmckay, @askhade: let us know what is your opinion.

Reply:
Would it be a bit cleaner if: 
  - the RewriteRule class ctor took the op type info
  - make RewriteRule::SatisfyCondition a non-virtual method, and add a private virtual method called SatisfyConditionImpl
  - the RewriteRule::SatisfyCondition implementation can do the op type checks and call SatisfyConditionImpl
  - derived classes pass one or more op types to the base RewriteRule ctor, and just implement SatisfyConditionImpl

Still doing the same thing just in different places, so there's no big difference between the two to me. 


---
In reply to: [268380878](https://github.com/Microsoft/onnxruntime/pull/670#discussion_r268380878) [](ancestors = 268380878)

Reply:
First, for OpTypeCondition, there is an implicit assumption that every re-write rule will be based on op type, which is not guaranteed. Second, due to OpTypeCondition, one more API AdditionalConditions is added. A new contributor will be puzzled when he/she implements this function. So, why not just have one SatisfyCondition for rewrite rule condition check?

Reply:
We are probably trying to fix too many things in this PR. 
So I kept a single SatisfyCondition in my latest commit.
But this has to be revisited.
Simplicity is one thing (and I am very much in favor of it), but we should also keep the rewrite rules application efficient and structured. And with the current design we are not doing a good job at it, which will become a problem as we add more rules.